### PR TITLE
Refactors ruby layer

### PR DESCRIPTION
**Pull request by adam-stokes**
- Compatibility with previous layer retained
- Will install via snapstore if snap-version is set and no ruby-version set,
  otherwise ruby-version takes precedence
- Add travis-ci, linting, black formatting
- Removed http/s-proxy from config as commands will now inherit from host os
  environment variables

Signed-off-by: Adam Stokes <battlemidget@users.noreply.github.com>

### CODE REVIEW

**Comment by adam-stokes** [comment]
ref #15, @mthaddon do you have a charm I can include in this repo (maybe the scm repo for cs:~mthaddon/discourse) for integration testing? Would also like a review on this PR

**Comment by mthaddon** [comment]
Per discussions IRL, probably best to use the ubuntu charm and install ruby as an integration test.

File: lib/charms/layer/ruby.py [hunk 1]
```diff
@@ -1,111 +1,76 @@
...
-    sh = shell(shell_cmd)
-    if sh.code > 0:
-        hookenv.status_set('blocked',
-                           'Problem with Ruby: {}'.format(sh.errors()))
+    is_snap_install = config.get("ruby-snap", None)
+    ruby_install_version = config.get("ruby-install-version", None)
+    if is_snap_install and not ruby_install_version:
+        sh.snap.install("ruby --channel {} --classic".format(config.get("ruby-snap")))
+        hookenv.status_set("maintenance", "Ruby installed from snapstore.")
+        return
```
Comment (stub42) [inline lib/charms/layer/ruby.py]: You need to catch 'not is_snap_install and not ruby-install-version' here, or the later steps will explode if someone set ruby_install_version to the empty string (expecting a default or latest version or similar).

File: lib/charms/layer/ruby.py [hunk 2]
```diff
@@ -1,111 +1,76 @@
...
-    hookenv.status_set('maintenance', 'Installing Ruby completed.')
-
-
-def download_ruby():
-    """ Downloads ruby tarball, puts it in /tmp
-    """
-    url = os.path.join(config['ruby-mirror'],
-                       'ruby-{}.tar.gz'.format(config['ruby-version']))
-    if not tarball_exists(url):
+    version = config.get("ruby-version", "2.6.4")
```
Comment (stub42) [inline lib/charms/layer/ruby.py]: This doesn't work as you expect, as the default will never be used. config['ruby-version'] always contains a value (even if None or the default '').

Instead:
version = config['ruby-version'] or '2.6.4'

**Review by stub42 — CHANGES_REQUESTED** [review]
This looks good.

I don't know what our requirements are here, but consider making the snap install the default, and consider dropping the non-snap install entirely. Ideally this implements best practice, and ideally there is e is only one best practice rather than two bestish practices.

**Comment by adam-stokes** [comment]
I would prefer to just stick with snaps, @mthaddon any objections? And if so would layer-snap cover our needs with some helpers wrapped around bundler and gem?

**Comment by mthaddon** [comment]
I don't have any objections to snaps only.

I don't know enough about the details to know if layer-snap with some helpers wrapped around bundler and gem would work, but it's quite possible it might.

**Comment by adam-stokes** [comment]
Cool, ill update the PR

**Review by stub42 — APPROVED** [review]
This all looks good to me.

While this is a backwards incompatible change, I don't think many charms will be affected and think that in this case it is better to tear the band-aid off quickly.

File: layer.yaml [hunk 3]
```diff
@@ -1,20 +1,7 @@
...
-      - zlib1g-dev
-      - libgdbm-dev
-      - openssl
-      - libicu-dev
-      - cmake
-      - pkg-config
-      - libxml2-dev
-      - libncurses5-dev
+      - wget
+      - git
```
Comment (stub42) [inline layer.yaml]: Do you still need wget and/or git ?

**Pull request merged**